src/cache/intern.rs

```diff
@@ -0,0 +1,72 @@
+use cfg_if::cfg_if;
+
+
+cfg_if! {
+    if #[cfg(feature = "enable-interning")] {
+        use std::thread_local;
+        use std::string::String;
+        use std::borrow::ToOwned;
+        use std::cell::RefCell;
+        use crate::JsValue;
+        use uluru::{LRUCache, Entry};
+
+
+        struct Pair {
+            key: String,
+            value: JsValue,
+        }
+
+        // TODO figure out a good default capacity
+        type Entries = LRUCache::<[Entry<Pair>; 1_024]>;
+
+        struct Cache {
+            entries: RefCell<Entries>,
+        }
+
+        thread_local! {
+            static CACHE: Cache = Cache {
+                entries: RefCell::new(LRUCache::default()),
+            };
+        }
+
+        fn get_js_string<'a>(cache: &'a mut Entries, key: &str) -> Option<&'a JsValue> {
+            cache.find(|p| p.key == key).map(|x| &x.value)
```

Comment:
Is this doing a linear search over all entries in the cache each time? I'm actually sort of surprised that's faster than copying the string across the boundary...

Reply:
Yes, but there's three things to keep in mind: the entries are kept in "most recently used" order, so more recently used strings will be found faster. And once it finds a match it immediately stops.

Secondly, the cache is quite small (only 1,024 items), which makes linear search fast. BTreeMap also uses linear search for small node sizes, because that's actually faster than doing something fancy.

Thirdly, doing a full `O(n)` copy of the string *and* encoding from UTF-8 to UTF-16 is *quite* slow (as shown in the real world benchmarks). If sending strings from Rust to JS was fast enough then I wouldn't have created an interning system in the first place.

That's all implementation stuff though, we can always optimize it to be faster later.

Reply:
My worry isn't so much for cache hits but cache misses. What happens if 1024 kilobyte strings are interned and then you pass `"a"` along the boundary? Every single string needs to be checked every single time for strings which aren't hits in the cache.

Like I agree encoding is slow, but this sort of interning is still slower than it would otherwise be because it's doing some linear lookup at least with respect to the length of the string, and this is also with respect to the cache.

My main point is that this seems like a case where we want some sort of LRU hash table, not just an LRU list.

Reply:
String equality *also* short-circuits if it doesn't match. So if you pass `"a"`, it will check the length of `"a"` and the cached string, and they won't match, so it'll immediately move onto the next string.

That means you have exactly 1,024 very fast operations (just comparing string lengths) before you get a cache miss.

And even if the length of the strings are the same, it will also abort as soon as even a single char is different.

So the only time it would be an issue is if you're passing a lot of strings which are prefixes of cached strings, and the prefix is really long.

I have no problem with improving the performance, but I think it should be done in a new PR, with proper benchmarks to make sure it doesn't reduce the performance.

(Also keep in mind that all of this only matters if you enable the `enable-interning` feature, if you don't enable that feature then there's 0 cost, the cache doesn't even exist at all).